fvm/transaction_test.go

```diff
@@ -222,6 +226,75 @@ func TestSafetyCheck(t *testing.T) {
 		require.NotContains(t, buffer.String(), "codes")
 		require.Equal(t, 0, proc.Retried)
 	})
+
+	t.Run("retriable errors causes retry", func(t *testing.T) {
+
+		rt := &ErrorReturningRuntime{TxErrors: []error{
+			runtime.Error{
+				Err: &runtime.ParsingCheckingError{
+					Err: &sema.CheckerError{
+						Errors: []error{
+							&sema.AlwaysFailingNonResourceCastingTypeError{
+								ValueType:  &sema.AnyType{},
+								TargetType: &sema.AnyType{},
+							}, // some dummy error
+							&sema.ImportedProgramError{
+								CheckerError: &sema.CheckerError{},
+								Location:     common.AddressLocation{Address: common.BytesToAddress([]byte{1, 2, 3, 4})},
+							},
+						},
+					},
+				},
+				Location: common.TransactionLocation{'t', 'x'},
+				Codes:    nil,
+				Programs: nil,
+			},
+			nil,
+		}}
+
+		log := zerolog.Nop()
+		txInvocator := NewTransactionInvocator(log)
+
+		vm := New(rt)
+		code := `doesn't matter`
+
+		proc := Transaction(&flow.TransactionBody{Script: []byte(code)}, 0)
+
+		ledger := state.NewMapLedger()
+		header := unittest.BlockHeaderFixture()
+		context := NewContext(log, WithBlockHeader(&header))
+
+		st := state.NewState(ledger)
+
+		err := txInvocator.Process(vm, context, proc, st)
+		assert.NoError(t, err)
+
+		require.Equal(t, 1, proc.Retried)
```

Comment:
how does it work that it only retries once, not twice? 
how does it work that the first try fails and retry worked?

Reply:
> how does it work that it only retries once, not twice?

Because second execution works fine

>how does it work that the first try fails and retry worked?

Not sure I understand, but there is a mocked up execution environment that returns error on first call, and `nil` on the second